modules/tools/open_space_visualization/auto_param_tuning.py

```diff
@@ -0,0 +1,118 @@
+from google.protobuf.internal import encoder
+import hybrid_a_star_visualizer
+import common.proto_utils as proto_utils
+from modules.planning.proto import planner_open_space_config_pb2
+
+
+def load_open_space_protobuf(filename):
+    open_space_params = planner_open_space_config_pb2.PlannerOpenSpaceConfig()
+    proto_utils.get_pb_from_text_file(filename, open_space_params)
+    return open_space_params
+
+
+def GetParamsForTunning(tunning_object):
+    param_names_and_range = []
+    if tunning_object == "coarse_trajectory":
+        param_names_and_range.append(
+            ("warm_start_config.traj_forward_penalty", 2.0))
+        param_names_and_range.append(
+            ("warm_start_config.traj_back_penalty", 2.0))
+        param_names_and_range.append(
+            ("warm_start_config.traj_gear_switch_penalty", 2.0))
+        param_names_and_range.append(
+            ("warm_start_config.traj_steer_penalty", 2.0))
+        param_names_and_range.append(
+            ("warm_start_config.traj_steer_change_penalty", 2.0))
+    return param_names_and_range
+
+
+def RandSampling(param_names_and_range, origin_open_space_params):
+    rand_num = 10
+    params_lists = []
+    for iter in range(0, rand_num):
+        rand_params = planner_open_space_config_pb2.PlannerOpenSpaceConfig()
+        rand_params.CopyFrom(origin_open_space_params)
+        for param in param_names_and_range:
+            exec("rand_params." +
+                 str(param[0]) + "=random.uniform(rand_params." +
+                 str(param[0])
+                 + " - " + str(param[1])
+                 + " ,rand_params." + str(param[0]) + " + " + str(param[1]) + ")")
+            params_lists.append(rand_params)
+    return params_lists
+
+
+def TestingParams(params_lists, tunning_object):
+    key_to_evaluations = {}
+    for iter in range(0, len(params_lists)):
+        evaluation = ParamEvaluation(params_lists[iter], tunning_object)
+        key_to_evaluations[iter] = evaluation
+    return key_to_evaluations
+
+
+def ParamEvaluation(params, tunning_object):
+    proto_utils.write_pb_to_text_file(params, original_file_path)
+    if tunning_object == "coarse_trajectory":
+        visualize_flag = False
+        success, x_out, y_out, phi_out, v_out, a_out, steer_out, planning_time = hybrid_a_star_visualizer.HybridAStarPlan(
+            visualize_flag)
+        if not success:
+            return float('inf')
+        else:
+            return planning_time
+
+
+def GetOptimalParams(params_lists, key_to_evaluations):
+    tmp = []
+    for key, value in key_to_evaluations.items():
+        tmptuple = (value, key)
+        tmp.append(tmptuple)
+
+    tmp = sorted(tmp)
+    optimal_params = params_lists[tmp[0][1]]
+    optimal_evaluation = tmp[0][0]
+    return optimal_params, optimal_evaluation
+
+
+if __name__ == '__main__':
+    random.seed(666)
+    original_file_path = "/apollo/modules/planning/conf/planner_open_space_config.pb.txt"
```

Comment:
Expand to take the original file path and destination file path as input argument as well for better integration in future.

Reply:
Done